nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/flow/synchronization/StandardVersionedComponentSynchronizer.java

```diff
@@ -1685,10 +1685,19 @@ public void synchronize(final ParameterContext parameterContext, final Versioned
 
     protected Set<String> getUpdatedParameterNames(final ParameterContext parameterContext, final VersionedParameterContext proposed) {
         final Map<String, ParameterValueAndReferences> originalValues = new HashMap<>();
-        parameterContext.getParameters().values().forEach(param -> originalValues.put(param.getDescriptor().getName(), getValueAndReferences(param)));
+        parameterContext.getEffectiveParameters().forEach((pd, param) -> originalValues.put(pd.getName(), getValueAndReferences(param)));
 
         final Map<String, ParameterValueAndReferences> proposedValues = new HashMap<>();
         if (proposed != null) {
+            if (proposed.getInheritedParameterContexts() != null) {
```

Comment:
With the changes above this could be made somewhat more functional:
```
            Optional.ofNullable(proposed.getInheritedParameterContexts())
                    .map(List::reversed)
                    .ifPresent(x -> x.forEach(name -> Optional.ofNullable(getParameterContextByName(name))
                            .ifPresent(inheritedContext -> collectValueAndReferences(inheritedContext, proposedValues))));
```
What do you think? Just an idea on the other hand list reverting won't apply cleanly to the 1.x branch so maybe it is cleaner to leave it as is. I will let you decide please let me know. Otherwise I'm +1, thanks for the test explaining the functionality.

Reply:
We should very much avoid trying to make things functional. We should always prefer Imperative unless there is a strong reason to use functional. I.e., in cases where the imperative is verbose but functional is trivial, such as:
```
final List<String> filtered = values.stream().filter(s -> s.contains("hello")).toList();
```
can be used over:
```
final List<String> filtered = new ArrayList<>();
for (final String value : values) {
  if (value.contains("hello") {
     filtered.add(value);
  }
}
```
Because it is simpler to understand the code and the intent. However, this code as-is, in its imperative form, is much simpler than the functional form.

Reply:
I will leave it as is, in keeping with @markap14's suggestion.

Reply:
Understood, it is ok from my side. I find functional code to be much easier to read and understand as it shows "what" is happening in contrast of imperative where it is more about "how" we achive our goal. I also agree that there are cases where functional approach could be more involved. In the end it is subjective I suppose. If the community prefers imperative then code consistency is more valuable.